spdxlib/documents_test.go

```diff
@@ -181,3 +182,89 @@ func Test2_2InvalidDocumentFailsValidation(t *testing.T) {
 		t.Fatalf("expected non-nil error, got nil")
 	}
 }
+
+// ===== 2.3 tests =====
+
+func Test2_3ValidDocumentPassesValidation(t *testing.T) {
+	// set up document and some packages and relationships
+	doc := &v2_3.Document{
+		SPDXVersion:    "SPDX-2.3",
+		DataLicense:    "CC0-1.0",
+		SPDXIdentifier: common.ElementID("DOCUMENT"),
+		CreationInfo:   &v2_3.CreationInfo{},
+		Packages: []*v2_3.Package{
+			{PackageName: "pkg1", PackageSPDXIdentifier: "p1"},
+			{PackageName: "pkg2", PackageSPDXIdentifier: "p2"},
+			{PackageName: "pkg3", PackageSPDXIdentifier: "p3"},
+			{PackageName: "pkg4", PackageSPDXIdentifier: "p4"},
+			{PackageName: "pkg5", PackageSPDXIdentifier: "p5"},
+		},
+		Relationships: []*v2_3.Relationship{
+			{
+				RefA:         common.MakeDocElementID("", "DOCUMENT"),
+				RefB:         common.MakeDocElementID("", "p1"),
+				Relationship: "DESCRIBES",
+			},
+			{
+				RefA:         common.MakeDocElementID("", "DOCUMENT"),
+				RefB:         common.MakeDocElementID("", "p5"),
+				Relationship: "DESCRIBES",
+			},
+			// inverse relationship -- should also get detected
+			{
+				RefA:         common.MakeDocElementID("", "p4"),
+				RefB:         common.MakeDocElementID("", "DOCUMENT"),
+				Relationship: "DESCRIBED_BY",
+			},
+			// different relationship
+			{
+				RefA:         common.MakeDocElementID("", "p1"),
+				RefB:         common.MakeDocElementID("", "p2"),
+				Relationship: "DEPENDS_ON",
+			},
+		},
+	}
+
+	err := ValidateDocument2_3(doc)
+	if err != nil {
+		t.Fatalf("expected nil error, got: %s", err.Error())
+	}
+}
+
+func Test2_3InvalidDocumentFailsValidation(t *testing.T) {
+	// set up document and some packages and relationships
+	doc := &v2_3.Document{
+		SPDXVersion:    "SPDX-2.1",
```

Comment:
is the old version intended for failure here?

Reply:
This was copied from the 2.2 version, but doesn't actually do much to test the failure condition; so I've updated it.